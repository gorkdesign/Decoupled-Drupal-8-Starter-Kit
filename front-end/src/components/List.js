import React, { Component } from 'react';
import PropTypes, { arrayOf, shape } from 'prop-types';
import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import MdAdd from 'react-icons/lib/md/add';

import ARTICLE_SHAPE from '../utils/articlePropType';
import Card from './Card';

const Fade = ({ children, ...props }) => (
  <CSSTransition
    {...props}
    timeout={1000}
    classNames="fade"
  >
    {children}
  </CSSTransition>
);
Fade.propTypes = {
  children: PropTypes.node.isRequired,
};

const ListItems = ({ articles, selectHandler, deleteHandler }) => articles.map(page => (
  <Fade duration={1000} key={page.nid} timeout={{ enter: 0, exit: 1000 }}>
    <Card
      page={page}
      selectHandler={selectHandler}
      deleteHandler={(event) => { deleteHandler(event, page.nid); }}
    />
  </Fade>
));

class List extends Component {
  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.articles.length !== prevState.articles.length) {
      return {
        articles: nextProps.articles,
      };
    }
    return null;
  }

  state = {
    articles: this.props.articles,
  }

  handleSearchInput = (target) => {
    const searchTerm = target.value;

    const { articles } = this.props;
    const searchedArticles = articles.map((page) => {
      const title = page.title.toLowerCase();
      const searchTermLower = searchTerm.toLowerCase();
      return title.includes(searchTermLower) ? page : null;
    });
    const filteredArticles = searchedArticles.filter(n => n != null);

    if (filteredArticles.length > 0) {
      this.setState({
        articles: filteredArticles,
      });
    } else if (target.value === '') {
      this.setState({
        articles: this.props.articles,
      });
    } else {
      this.setState({
        articles: [],
      });
    }
  }

  render() {
    const {
      addHandler, isModalVisible, onDeleteModalToggle, onDeleteModalOk, selectHandler, deleteHandler,
    } = this.props;

    return (
      <div className="">
        <div className="container">
          <div
            role="button"
            tabIndex={0}
            onKeyUp={() => { addHandler(); }}
            className="py-3"
            onClick={() => { addHandler(); }}
          >
            <div className="add">
              <MdAdd />
              <h2 className="card-title">Add</h2>
            </div>
          </div>

          <div className="input-group input-group-lg">
            <div className="input-group-prepend">
              <span className="input-group-text" id="inputGroup-sizing-lg">Search By Title</span>
            </div>
            <input onChange={({ target }) => this.handleSearchInput(target)} placeholder="Content Title..." type="text" className="form-control search-bar" aria-label="Large" aria-describedby="inputGroup-sizing-sm" />
          </div>

          {
            this.state.articles.length > 0 ?
              <TransitionGroup className="item-list">
                <ListItems
                  articles={this.state.articles}
                  selectHandler={selectHandler}
                  deleteHandler={deleteHandler}
                />
              </TransitionGroup>
          : <div className="no-result-text h1 text-center">No Results Found</div>
          }

          <Modal isOpen={isModalVisible} toggle={onDeleteModalToggle} backdrop>
            <ModalHeader toggle={onDeleteModalToggle}>Confirmation</ModalHeader>
            <ModalBody>
              Are you sure you want to remove this?
            </ModalBody>
            <ModalFooter>
              <Button color="secondary" onClick={onDeleteModalToggle}>Cancel</Button>
              {' '}
              <Button color="primary" onClick={onDeleteModalOk}>Delete</Button>
            </ModalFooter>
          </Modal>
        </div>
      </div>
    );
  }
}

List.propTypes = {
  articles: arrayOf(shape(ARTICLE_SHAPE).isRequired).isRequired,
  isModalVisible: PropTypes.bool.isRequired,
  addHandler: PropTypes.func.isRequired,
  onDeleteModalToggle: PropTypes.func.isRequired,
  onDeleteModalOk: PropTypes.func.isRequired,
};

export default List;
