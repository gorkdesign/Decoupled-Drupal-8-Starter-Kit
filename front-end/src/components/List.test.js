import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import List from './List';

const pageModel = {
  author: {
    name: 'test',
  },
  body: {
    value: 'body test value',
  },
  images: [
    {
      entity: {
        image: {
          derivative: {
            url: 'http://d8d.loc/test.jpg',
            mid: 1,
          },
        },
      },
    },
  ],
  nid: 34,
  title: 'TEST TITLE',
  uuid: '09ba8da5-e08d-4972-a22b-ad7bb0a44d7f',
};

const articles = [];
for (let i = 0; i < 10; i += 1) {
  articles.push({
    ...pageModel,
    author: { name: `test${i}` },
    nid: pageModel.nid + i,
    title: `${pageModel.title} ${i}`,
    uuid: pageModel.uuid + i,
  });
}

const initialPropsState = {
  isAuthenticated: true,
  isLoggingIn: false,
  isModalVisible: false,
  addArticleHandler: jest.fn(),
  selectArticleHandler: jest.fn(),
  onDeleteModalToggle: jest.fn(),
  onDeleteModalOk: jest.fn(),
  articles,
};

describe('List', () => {
  describe('Snapshots', () => {
    const output = shallow(<List {...initialPropsState} />);

    it('should render correctly', () => {
      expect(shallowToJson(output)).toMatchSnapshot();
    });

    it('should render a list of articles as a Card', () => {
      expect(output.find('Card').length).toEqual(articles.length);
    });
  });
});
