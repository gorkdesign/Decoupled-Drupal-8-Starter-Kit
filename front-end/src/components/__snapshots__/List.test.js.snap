// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Snapshots should render correctly 1`] = `
<div
  className=""
>
  <div
    className="container"
  >
    <div
      className="py-3"
      onClick={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="add"
      >
        <MdAdd />
        <h2
          className="card-title"
        >
          Add
        </h2>
      </div>
    </div>
    <TransitionGroup
      childFactory={[Function]}
      className="item-list"
      component="div"
    >
      <Fade
        duration={1000}
        key="34"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test0",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 34,
              "title": "TEST TITLE 0",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f0",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
      <Fade
        duration={1000}
        key="35"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test1",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 35,
              "title": "TEST TITLE 1",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f1",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
      <Fade
        duration={1000}
        key="36"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test2",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 36,
              "title": "TEST TITLE 2",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f2",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
      <Fade
        duration={1000}
        key="37"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test3",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 37,
              "title": "TEST TITLE 3",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f3",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
      <Fade
        duration={1000}
        key="38"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test4",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 38,
              "title": "TEST TITLE 4",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f4",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
      <Fade
        duration={1000}
        key="39"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test5",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 39,
              "title": "TEST TITLE 5",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f5",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
      <Fade
        duration={1000}
        key="40"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test6",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 40,
              "title": "TEST TITLE 6",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f6",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
      <Fade
        duration={1000}
        key="41"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test7",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 41,
              "title": "TEST TITLE 7",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f7",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
      <Fade
        duration={1000}
        key="42"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test8",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 42,
              "title": "TEST TITLE 8",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f8",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
      <Fade
        duration={1000}
        key="43"
        timeout={
          Object {
            "enter": 0,
            "exit": 1000,
          }
        }
      >
        <Card
          deleteHandler={[Function]}
          page={
            Object {
              "author": Object {
                "name": "test9",
              },
              "body": Object {
                "value": "body test value",
              },
              "images": Array [
                Object {
                  "entity": Object {
                    "image": Object {
                      "derivative": Object {
                        "mid": 1,
                        "url": "http://d8d.loc/test.jpg",
                      },
                    },
                  },
                },
              ],
              "nid": 43,
              "title": "TEST TITLE 9",
              "uuid": "09ba8da5-e08d-4972-a22b-ad7bb0a44d7f9",
            }
          }
          selectArticleHandler={[Function]}
        />
      </Fade>
    </TransitionGroup>
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransition={
        Object {
          "mountOnEnter": true,
          "timeout": 150,
        }
      }
      fade={true}
      isOpen={false}
      keyboard={true}
      modalTransition={
        Object {
          "timeout": 300,
        }
      }
      onClosed={[Function]}
      onOpened={[Function]}
      role="dialog"
      toggle={[Function]}
      zIndex={1050}
    >
      <ModalHeader
        closeAriaLabel="Close"
        tag="h5"
        toggle={[Function]}
        wrapTag="div"
      >
        Confirmation
      </ModalHeader>
      <ModalBody
        tag="div"
      >
        Are you sure you want to remove this?
      </ModalBody>
      <ModalFooter
        tag="div"
      >
        <Button
          color="secondary"
          onClick={[Function]}
          tag="button"
        >
          Cancel
        </Button>
         
        <Button
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Delete
        </Button>
      </ModalFooter>
    </Modal>
  </div>
</div>
`;
